components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java

```diff
@@ -221,6 +222,28 @@ public void unregisterHookSupplier(final Supplier<StorageTxHook> hookSupplier) {
     hookSuppliers.remove(hookSupplier);
   }
 
+  private static final int MAX_CONCURRENT_MODIFICATION_ATTEMPTS = 3;
+
+  @Override
+  public <T> T perform(final Operation<T> operation) {
+    checkNotNull(operation);
+    OConcurrentModificationException lastConcurrentException = null;
+    for (int attempt = 0; attempt < MAX_CONCURRENT_MODIFICATION_ATTEMPTS; attempt++) {
+      try (StorageTx tx = openTx()) {
+        try {
+          T result = operation.execute(tx);
+          tx.commit();
+          return result;
+        }
+        catch (OConcurrentModificationException e) {
+          lastConcurrentException = e;
+          log.debug("Operation {} attempt failed, retrying: {}", attempt + 1, operation, e);
```

Comment:
The operations added in this PR are anonymous classes so you might want to add `toString` to those implementations (rather than have to guess which class is `$2` for example)